pycloudlib/azure/cloud.py

```diff
@@ -937,11 +948,32 @@ def snapshot(
 
         return image_id
 
-    def delete_resource_group(self):
-        """Delete a resource group."""
-        if self.resource_group:
-            self.resource_client.resource_groups.begin_delete(
-                resource_group_name=self.resource_group.name
-            )
+    def delete_resource_group(self, resource_group_name: Optional[str] = None):
+        """Delete a resource group.
 
+        If no resource group is provided, delete self.resource_group
+        """
+        if resource_group_name is None and self.resource_group:
+            resource_group_name = self.resource_group.name
             self.resource_group = None
+        if resource_group_name:
+            with contextlib.suppress(ResourceNotFoundError):
+                self.resource_client.resource_groups.begin_delete(
+                    resource_group_name=resource_group_name
+                )
+
+    # pylint: disable=broad-except
+    def clean(self) -> List[Exception]:
+        """Cleanup ALL artifacts associated with this Cloud instance.
+
+        This includes all instances, snapshots, resources, etc.
+        To ensure cleanup isn't interrupted, any exceptions raised during
+        cleanup operations will be collected and returned.
+        """
+        exceptions = super().clean()
+        for resource_group in self.created_resource_groups:
```

Comment:
Some of this is a bit tacked on as I didn't want to do broader class refactors at the same time. If you look at `_create_resource_group()`, if the resource group under the given tag already exists, we'll use it, otherwise we'll create it and add it to `self.created_resource_groups` while also making it `self.resource_group`, so I don't think clean will do nothing here unless the resource group already existed.